File: Android.Conference.WebRTC/app/src/main/java/fm/android/conference/webrtc/LocalMedia.java
package fm.android.conference.webrtc;

import java.util.Locale;

import android.view.*;
import fm.*;
import fm.icelink.webrtc.*;

public class LocalMedia
{
    // We're going to need both audio and video
    // for this example. We can constrain the
    // video slightly for performance benefits.
    private boolean audio = true;
    private boolean video = true;
    private int videoWidth     = 640;
    private int videoHeight    = 480;
    private int videoFrameRate = 15;
    
    private LocalMediaStream localMediaStream;
    public  LocalMediaStream getLocalMediaStream() { return localMediaStream; }

    private AndroidLayoutManager layoutManager;
    public  AndroidLayoutManager getLayoutManager() { return layoutManager; }
    
    public void start(final ViewGroup videoContainer, final SingleAction<String> callback) throws Exception
    {
        // WebRTC audio and video streams require us to first get access to
        // the local media (microphone, camera, or both).
        UserMedia.getMedia(new GetMediaArgs(audio, video)
        {{
            // Uncomment this line to do in-app screen capture.
            //setVideoCaptureProvider(new AndroidScreenCaptureProvider(videoContainer.getRootView()));
            
            // Uncomment this line to do full-screen capture (Android 5.0+ only, requires the calling Activity to pass in a MediaProjection instance).
            // You will likely also want to comment out "app.pauseLocalVideo();" and "app.resumeLocalVideo();" in VideoActivity.java to prevent the
            // media capture from being paused/resumed when the activity is paused/resumed.
            //setVideoCaptureProvider(new AndroidMediaProjectionCaptureProvider(projection)); 
            
            setVideoWidth(videoWidth);         // optional
            setVideoHeight(videoHeight);       // optional
            setVideoFrameRate(videoFrameRate); // optional
            setDefaultVideoPreviewScale(LayoutScale.Contain); // optional
            setDefaultVideoScale(LayoutScale.Contain);        // optional
            setOnFailure(new SingleAction<GetMediaFailureArgs>()
            {
                public void invoke(GetMediaFailureArgs e)
                {
                    callback.invoke(String.format(Locale.getDefault(), "Could not get media. %s", e.getException().getMessage()));
                }
            });
            setOnSuccess(new SingleAction<GetMediaSuccessArgs>()
            {
                public void invoke(GetMediaSuccessArgs e)
                {
                    try
                    {
                        // We have successfully acquired access to the local
                        // audio/video device! Grab a reference to the media.
                        // Internally, it maintains access to the local audio
                        // and video feeds coming from the device hardware.
                        localMediaStream = e.getLocalStream();

                        // This is our local video control, a View. It is
                        // constantly updated with our live video feed since
                        // we requested video above. Add it directly to the UI
                        // or use the IceLink layout manager, which we do below.
                        View localVideoControl = (View)e.getLocalVideoControl();

                        // Create an IceLink layout manager, which makes the task
                        // of arranging video controls easy. Give it a reference
                        // to a View that can be filled with video feeds.
                        layoutManager = new AndroidLayoutManager(videoContainer);

                        // Position and display the local video control on-screen
                        // by passing it to the layout manager created above.
                        layoutManager.setLocalVideoControl(localVideoControl);
    
                        callback.invoke(null);
                    }
                    catch (Exception ex)
                    {
                        ex.printStackTrace();
                    }
                }
            });
        }});
    }
    
    public void stop(final SingleAction<String> callback) throws Exception
    {
        // Clear out the layout manager.
        if (layoutManager != null)
        {
            layoutManager.removeRemoteVideoControls();
            layoutManager.unsetLocalVideoControl();
            layoutManager = null;
        }

        // Stop the local media stream.
        if (localMediaStream != null)
        {
            localMediaStream.stop();
            localMediaStream = null;
        }

        callback.invoke(null);
    }
}

File: ios/LocalMedia.m
//
//  LocalMedia.m
//  iOS.Conference.WebRTC
//
//  Created by Adrian Mercado on 2014-06-09.
//  Copyright (c) 2014 Frozen Mountain Software. All rights reserved.
//

#import "LocalMedia.h"

@implementation LocalMedia

// We're going to use both audio and video
// for this example.
static bool audio          = YES;
static bool video          = YES;
static int  videoFrameRate = 15;

@synthesize layoutManager = _layoutManager;
@synthesize localMediaStream = _localMediaStream;

- (void)start:(UIView *)videoContainer callback:(void (^)(NSString *))callback
{
  [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord
                                   withOptions:
   AVAudioSessionCategoryOptionAllowBluetooth |
   AVAudioSessionCategoryOptionDefaultToSpeaker error:nil];

  // WebRTC audio and video streams require us to first get access to
  // the local media (microphone, camera, or both).
  FMIceLinkWebRTCGetMediaArgs *getMediaArgs = [FMIceLinkWebRTCGetMediaArgs getMediaArgsWithAudio:audio video:video];
  //[getMediaArgs setVideoCaptureProvider:[[[FMIceLinkWebRTCScreenCaptureProvider alloc] init] autorelease]]; // uncomment this line to do in-app screen capture
  [getMediaArgs setVideoPreset:AVCaptureSessionPresetMedium]; // optional
  [getMediaArgs setVideoFrameRate:videoFrameRate];            // optional
  [getMediaArgs setDefaultVideoPreviewScale:FMIceLinkWebRTCLayoutScaleContain]; // optional
  [getMediaArgs setDefaultVideoScale:FMIceLinkWebRTCLayoutScaleContain];        // optional
  [getMediaArgs setOnFailureBlock:^(FMIceLinkWebRTCGetMediaFailureArgs *e)
   {
     callback([NSString stringWithFormat:@"Could not get media. %@", e.exception.message]);
   }];
  [getMediaArgs setOnSuccessBlock:^(FMIceLinkWebRTCGetMediaSuccessArgs *e)
   {
     // We have successfully acquired access to the local
     // audio/video device! Grab a reference to the media.
     // Internally, it maintains access to the local audio
     // and video feeds coming from the device hardware.
     _localMediaStream = e.localStream;

     // This is our local video control, a UIView (iOS) or
     // and NSView (Mac). It is constantly updated with our
     // live video feed since we requested video above. Add
     // it directly to the UI or use the IceLink layout manager,
     // which we do below.
     UIView *localVideoControl = (UIView *)e.localVideoControl;

     // Create an IceLink layout manager, which makes the task
     // of arranging video controls easy. Give it a reference
     // to a UIView that can be filled with video feeds.
     _layoutManager = [FMIceLinkWebRTCLayoutManager layoutManagerWithContainer:videoContainer];

     // Position and display the local video control on-screen
     // by passing it to the layout manager created above.
     [_layoutManager setLocalVideoControl:localVideoControl];

     callback(nil);
   }];
  [FMIceLinkWebRTCUserMedia getMediaWithArgs:getMediaArgs];
}

- (void)stop:(void (^)(NSString *))callback
{
  // Clear out the layout manager.
  [_layoutManager removeRemoteVideoControls];
  [_layoutManager unsetLocalVideoControl];
  _layoutManager = nil;

  // Stop the local media stream.
  [_localMediaStream stop];
  _localMediaStream = nil;

  callback(nil);
}

@end
